### PR TITLE
Use `readFully()` to read exactly the number of bytes we expect from the Stream

### PR DESCRIPTION
**Pull request by jkakavas**
Instead of using `read()` and checking that the bytes read are what we expect,
use `readFully()` which will read exactly the number of bytes
while keep reading until the end of the stream or throw an
`EOFException` if not all bytes can be read.

This approach keeps the simplicity of using CipherInputStream while
working as expected with all Security Providers (SunJCE, BouncyCastle, BouncyCastleFips )

~~Calls explicitly `doFinal()` on the byte array of plaintext and~~
~~ciphertext, instead of making use of `Cipher{Input,Output}Stream`.~~
~~While functionally equivelant, CipherInputStream merhod does not~~
~~work well with BouncyCastle (FIPS) Security Provider's implementation~~
~~of AES GCM as the byte array that is read from the `CipherInputStream`~~
~~backed `DataInputStream` is always some bytes short (17 vs 20 for~~
~~the case of `keystore.seed` value).~~

### CODE REVIEW

**Review by jaymode — COMMENTED** [review]
> While functionally equivelant, CipherInputStream merhod does not
work well with BouncyCastle (FIPS) Security Provider's implementation
of AES GCM as the byte array that is read from the CipherInputStream
backed DataInputStream is always some bytes short (17 vs 20 for
the case of keystore.seed value).

This sounds like it could be a bug in the BC implementation? Can you create a small reproducer and report it upstream. I am not a fan of adding workarounds like this unless we have a upstream issue so that we can plan to remove the workaround in the future.

If we're going to remove the use of CipherInputStream, we should add it to forbidden apis otherwise it will crop up somewhere else.

**Comment by rjernst** [comment]
+1 to what Jay said.

**Comment by jkakavas** [comment]
Thanks for the comments. I started a [discussion upstream](https://www.bouncycastle.org/devmailarchive/msg15555.html) and provided them with test cases for reproduction. 

> . I am not a fan of adding workarounds like this 

Apart from this approach being a workaround to the original version, do you see anything that is inherently wrong with using `doFinal()` explicitly ?

> If we're going to remove the use of CipherInputStream, we should add it to forbidden apis otherwise it will crop up somewhere else. 

It does seem to affect other AES modes,I was able to reproduce it with AES/CBC/PKCS7Padding too for instance.

**Comment by jaymode** [comment]
> I started a discussion upstream and provided them with test cases for reproduction.

Thank you.

> do you see anything that is inherently wrong with using doFinal() explicitly

There is nothing wrong with this aspect explicitly. In many instances the Cipher*Stream make things simpler, which is why I prefer their usage. Also, we should add comments in the code about why we do things a certain way with links to things like the upstream discussion so that the next person to come across this has a clue why things were done a certain way.

**Comment by jkakavas** [comment]
Based on the feedback from the upstream mailing list, it turns out this has to do with how the data is read from the stream, and thus affects only the `CipherInputStream`.  I changed to a different approach, using `readFully()` on the DataInputStream , that works as expected for both providers. I plan on renaming this PR in order to keep the history of the discussion, but I might as well close this and open a new one if you think it is cleaner.

File: server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 package org.elasticsearch.common.settings;
 
+
```
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: nit: unnecessary change?

File: server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java [hunk 2]
```diff
@@ -317,38 +319,33 @@ public void decrypt(char[] password) throws GeneralSecurityException, IOExceptio
              DataInputStream input = new DataInputStream(bytesStream)) {
```
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: I think we should add a test that would have failed using the old method. This would probably require a small refactoring to do this; I am thinking instead of `new ByteArrayInputStream` we call a method that could be overridden in a test. There we would have a special stream where a call to `read` would not fully populate the byte array

**Review by jaymode — CHANGES_REQUESTED** [review]
I think it is fine to keep this PR. Let's just update the title and description to be appropriate of the change. I'd like to have a test for this and don't feel like it should be too difficult to add.

File: server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java (hunk 2 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: I don't think we need to override any methods. We can form streams of bytes in whatever order we want in tests.
However, if I understand the problem correctly, it has to do with read boundaries. Eg, an input stream may choose to return less bytes to a `read` call than the array can contain, since it may just copy the rest of an existing buffer (eg a decrypted block in CipherInputStream). By calling `readFully`, multiple read calls are made to ensure the array is filled completely or EOF is reached. I think this will be very difficult to mimic, given the behavior of `read` here will be implementation dependent. The best we could do is add tests to trigger each of the possible EOFs from `readFully` calls by truncating the keystore at certain points (although this would probably be difficult to do within the encrypted bytes).

I do think we should have a test to ensure there cannot be garbage after the encrypted data.

Comment (jkakavas) [reply server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: Revisiting this after coming back. 
> I do think we should have a test to ensure there cannot be garbage after the encrypted data

Care to elaborate @rjernst ? You mean attempting to `readFully()` into a larger array and verifying that we get the EOFException rather than keep reading garbage ?

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: I mean adding a test with garbage at the end, so that we know the last readFully logic works. The intermediate readFully's are harder, as I described above.

Comment (jaymode) [reply server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: > I do think we should have a test to ensure there cannot be garbage after the encrypted data.

If I understand correctly, we should fail if there is garbage right? Currently this is not the case at least based on the test @jkakavas added.

Comment (jkakavas) [reply server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: `readFully` reads exactly the number of bytes we instruct it to, or throws an EOFException. I interpreted @rjernst comment above to mean that we need to have a test verifying that even if there  more garbage, readFully won't read them, but read the bytes it should(this is what the test I added does, and it should not fail ) 
Another way to interpret the comment I guess, would be to write less bytes to would be to write fewer bytes to the `DataOutputStream` than we indicate with the `writeInt` above and make sure that on decryption, readFully throws the EOFException instead of reading in garbage in order to fill our byte array - I can add that too.
Writing about this, the second interpretation makes much more sense I guess..

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: That was not my intention. I meant it the way @jaymode suggested: we should fail with garbage, and have a test for that.

**Comment by jkakavas** [comment]
Added check and test for trailing garbage. As discussed, we can address the slightly more complicated test case of cross buffer boundaries in a follow-up PR, so this is ready for another look @rjernst, @jaymode

**Review by jaymode — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java [hunk 3]
```diff
@@ -317,38 +318,36 @@ public void decrypt(char[] password) throws GeneralSecurityException, IOExceptio
...
-                throw new SecurityException("Keystore has been corrupted or tampered with");
-            }
+            input.readFully(iv);
             int encryptedLen = input.readInt();
             encryptedBytes = new byte[encryptedLen];
-            if (input.read(encryptedBytes) != encryptedLen) {
-                throw new SecurityException("Keystore has been corrupted or tampered with");
-            }
+            input.readFully(encryptedBytes);
+        } catch (EOFException e) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: While the exception message was the same in all cases before, we still would have been able to tell where the failed read happened (ie where in the bytes the corrupt/tamped data occurs). Are we sure we want to drop the root cause exception here? I think that at least warrants a comment here as to why.

File: server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java [hunk 4]
```diff
@@ -317,38 +318,36 @@ public void decrypt(char[] password) throws GeneralSecurityException, IOExceptio
...
-                    throw new SecurityException("Keystore has been corrupted or tampered with");
-                }
+                input.readFully(entryBytes);
                 entries.get().put(setting, new Entry(entryType, entryBytes));
             }
+            if (input.read() != -1) {
+                throw new SecurityException("Keystore has been corrupted or tampered with");
+            }
+        } catch (EOFException e) {
+            throw new SecurityException("Keystore has been corrupted or tampered with");
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: Same comment about dropping the root cause as above.

File: server/src/test/java/org/elasticsearch/common/settings/KeyStoreWrapperTests.java [hunk 5]
```diff
@@ -104,6 +107,57 @@ public void testUpgradeNoop() throws Exception {
         assertEquals(seed.toString(), keystore.getString(KeyStoreWrapper.SEED_SETTING.getKey()).toString());
     }
 
+    public void testFailWhenStreamNotConsumed() throws Exception {
```
Comment (rjernst) [inline server/src/test/java/org/elasticsearch/common/settings/KeyStoreWrapperTests.java]: This seems to test just one case. Can it be duplicated/refactored so it tests each of the readFully calls?

**Review by rjernst — APPROVED** [review]
LGTM, I left a couple final requests.

**Comment by jkakavas** [comment]
@rjernst I addressed your comments, I think this is ready (pending green CI), I would appreciate a :+1: before merging

**Review by rjernst — APPROVED** [review]
LGTM
